### PR TITLE
Performance: Remove unused models, reduce memory usage

### PR DESCRIPTION
**Pull request by bmtcril**
Based on the dbt lineage graph with Superset chart exposures, these models did not appear to be used. Some were moved back to Superset due to prewhere issues and never removed, some were simply cut as features. The materialized views were slowing down inserts and causing large memory spikes when trying to insert large batches of data, causing data load failures and eventually could have broken normal inserts as well.

Note: This may be a breaking change for consumers of the data API, it is suggested that consumers move these models to their own downstream dbt package if they are using any of the deleted models in custom reports.

This includes the changes from #110 and extends that pattern to `user_pii` to cut down on memory consumption. It also makes the cache lifetime of block names and course names configurable. This may be desirable on large instances with lots of courses, and cut down on the frequency of those potentially large memory consuming operations. The tradeoff would be longer wait times for new courses / blocks to appear in reports.

This is part 1 of at least 2 PRs for large scale performance optimization, and has been tested successfully on a fresh ClickHouse Cloud "production" cluster with the 300M row dataset with good performance.

TODO in this PR: 
- Test in tutor plugin repo
- Update release notes, or find a way to automatically drop the models we're removing

### CODE REVIEW

File: models/courses/most_recent_course_blocks.sql [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        schema=env_var("ASPECTS_EVENT_SINK_DATABASE", "event_sink"),
+        engine=get_engine("ReplacingMergeTree()"),
+        order_by="(location)",
+        post_hook="OPTIMIZE TABLE {{ this }} {{ on_cluster() }} FINAL",
+    )
+}}
+
+select
+    location,
+    display_name as block_name,
```
Comment (saraburns1) [inline models/courses/most_recent_course_blocks.sql:13]: something going on with this, still looking into it 
image

Comment (saraburns1) [reply models/courses/most_recent_course_blocks.sql:13]: fixed in https://github.com/openedx/tutor-contrib-aspects/pull/907

**Review by saraburns1 — APPROVED** [review]
Tested locally along with https://github.com/openedx/tutor-contrib-aspects/pull/907
Alembic, dbt, & superset dashboards all look good

**Comment by bmtcril** [comment]
@Ian2012 this should be good for review now

**Review by Ian2012 — APPROVED** [review]
LGTM, will you release this as v4?

**Comment by bmtcril** [comment]
I think we'll want to wait for the next breaking performance PR (in progress) and release them together as v4, since it will potentially be a messy upgrade path for people with customizations built on the v3 schema I'd like to do one big change instead of two big changes.

That said, we have some options for how to help people upgrade. Mostly what will matter are:

- Dropping unused models (optionally?)
- Model changes (renamed or columns added/removed/types changed)

For people just using out-of-the-box Aspects we could just give a command that they could run to drop the list of unused tables, but it's more complicated for people who have built custom datasets or reports. I *think* the best bet is to write the drop script and thoroughly document the upgrade process for people to move models into their own dbt project / know how the schema has changed to know what they need to update in their customizations. 

Do you think that would be sufficient?

**Comment by Ian2012** [comment]
I agree with the upgrade path and dropping the old models, but I'm worried about how that command should be written. Maybe we should:
- Drop all reporting models
- Drop non-standard tables on the database `xapi`. `event_sink` can be extended easily so we should only drop what we know.
- Add documentation that custom tables and reports should be on a different database which can be created by Aspects (`aspects_extensions`?)
- Make sure old models are always deleted (how?, no idea)

**Comment by bmtcril** [comment]
I think the command would have to be version specific or something written into the dbt package, but hard code the things we want to drop if they exist and just continually add to that. I don't love the idea of dropping unknown objects. 

I like having a specific schema for extensions to write to, then we can create it and set permissions so people don't need to worry about it. I have some concern that multiple extensions will try to have the same name, since not everyone will use dbt to write their schema extensions, so we should also probably have people prefix their object names to the package so they won't collide.

I think we can do a `dbt run-operation` to run the drop script as a dbt macro.

**Pull request merged**